### PR TITLE
fix: skip Content Loop/Carousel render for REST revisions and autosaves

### PR DESCRIPTION
**Pull request by claudiulodro**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guidelines](https://github.com/Automattic/newspack-workspace/blob/main/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

On sites with many posts, editing a post was slower than it should be. The revisions and autosaves REST endpoints render a post's content through `the_content`, which runs every dynamic block's render callback, so each Content Loop and Carousel block ran a full query for output the editor never displays. The editor autosaves on a timer, so this cost repeated throughout an editing session.

Both blocks now return early when a revision or autosave is being prepared for the REST API. A small helper tracks the REST route being served and reports when it is a revisions or autosaves request; normal REST requests and front-end rendering are unchanged.

This is a hotfix branched off `release`.

### How to test the changes in this Pull Request:

1. Add a Content Loop block and a Carousel block to a post and publish. Both render on the front end.
2. Request `wp/v2/posts/{id}/revisions/{rev}?context=edit` for that post. `content.rendered` omits both blocks.
3. Request `wp/v2/posts/{id}?context=edit`. `content.rendered` still includes both blocks.
4. Run `n test-php` in `newspack-blocks`. All tests pass, including the new `test-revision-render-skip.php`.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

The route is cleared on `rest_post_dispatch`, so the captured value cannot leak into later work in a long-running process.

### CODE REVIEW

File: plugins/newspack-blocks/includes/class-newspack-blocks.php [hunk 1]
```diff
@@ -42,6 +51,8 @@ public static function init() {
 		add_filter( 'the_content', [ __CLASS__, 'end_content_render_pass' ], PHP_INT_MAX );
 		add_filter( 'body_class', [ __CLASS__, 'add_body_classes' ] );
 		add_filter( 'admin_body_class', [ __CLASS__, 'add_body_classes' ] );
+		add_filter( 'rest_pre_dispatch', [ __CLASS__, 'capture_rest_route' ], 10, 3 );
+		add_filter( 'rest_post_dispatch', [ __CLASS__, 'clear_rest_route' ], 10, 3 );
```
Comment (chickenn00dle) [inline plugins/newspack-blocks/includes/class-newspack-blocks.php]: **Suggestion:** Captured route outlives in-process REST requests — `rest_do_request()` fires `rest_pre_dispatch` (inside `dispatch()`) but never `rest_post_dispatch`, so an in-process request to a revisions or autosaves route leaves `$current_rest_route` set, and every Content Loop and Carousel rendered later in that PHP process returns `''`. Could we bracket on `rest_request_before_callbacks` / `rest_request_after_callbacks` instead? They wrap the endpoint callback on every dispatch path (HTTP, `rest_do_request()`, batch, embeds, preload), and a small stack in place of the single static would also restore the outer route after a nested request. A test that calls `rest_do_request()` on a revisions route and then renders the block outside REST would lock it in.

<details><summary>Evidence</summary>

Verified on a test site: rendering a post with both blocks through `the_content` included them; after `rest_do_request( new WP_REST_Request( 'GET', '/wp/v2/posts/10/revisions' ) )` returned 200, the same render in the same process included neither. `rest_do_request()` is a bare `rest_get_server()->dispatch( $request )` (`rest-api.php:613-616`), and `rest_preload_api_request()` fires `rest_post_dispatch` only on a 200. The current tests don't see this because `tear_down()` fires `rest_post_dispatch` by hand.

It also reaches the page cache. With a `template_redirect` callback that calls `rest_do_request()` on `/wp/v2/posts/10/revisions` (a stand-in for third-party code), a logged-out request rendered neither block even though that request was refused with a 401: the route is captured before the permission check. Once the block cache was cold, Batcache stored the empty page and served it on the next request (`Batcache-Set`, then `Batcache-Hit`). Logged-in editors see it every time, since block caching is off for them.

I didn't find Newspack code that makes that kind of request, so as far as I can tell this doesn't happen today; it would take a third-party plugin, cron job or CLI command.

</details>

File: plugins/newspack-blocks/tests/test-revision-render-skip.php [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	 */
+	private function render_carousel() {
+		return do_blocks( '<!-- wp:newspack-blocks/carousel {"slidesPerView":1,"postsToShow":3} /-->' );
+	}
+
+	/**
+	 * The Content Loop must not run its query while a revision is being prepared
+	 * for the REST API.
+	 */
+	public function test_content_loop_skips_render_during_revisions_rest_request() {
```
Comment (chickenn00dle) [inline plugins/newspack-blocks/tests/test-revision-render-skip.php]: **Nit:** Route × block cases could share a data provider — The four skip cases and the two non-revision cases differ only in the block markup, the class marker and the route. A provider over `[ markup, marker, route, expect_rendered ]` would keep what each one pins down in fewer lines, and make an `/autosaves/{id}` row, which nothing covers yet, a one-line addition.

File: plugins/newspack-blocks/tests/test-revision-render-skip.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+			'wpnbha',
+			$this->render_content_loop(),
+			'The Content Loop must still render on a non-revision REST request.'
+		);
+	}
+
+	/**
+	 * A front-end render, with no REST request in flight at all, is untouched.
+	 */
+	public function test_content_loop_renders_on_the_front_end() {
```
Comment (chickenn00dle) [inline plugins/newspack-blocks/tests/test-revision-render-skip.php]: **Nit:** Front-end test passes with or without the change — `test_content_loop_renders_on_the_front_end` passes on the pre-change code, and `test_content_loop_renders_during_a_non_revision_rest_request` already fails if the skip ever becomes unconditional. It could become the `rest_do_request()` leak test from the comment on `class-newspack-blocks.php:55` instead.

**Review by chickenn00dle — APPROVED** [review]
Tested and this works great! Claude found a pretty specific edge case that doesn't seem reproducible with only our suite, so feel free to ignore the suggestion below (as well as the nits)!

Review scope: 2 of 3 reviewer passes ran; skipped: codex.

**Comment by chickenn00dle** [comment]
Its worth calling out another content loop block PR I've got up here: https://github.com/Automattic/newspack-workspace/pull/1110

This one targets `main` though. I wanted to let this go through alpha testing first since it touches such a critical part of our publishers' sites but it feels adjacent to this one so happy to make it a hotfix instead!
